ouroboros-consensus/docs/future-ledger-hd.md

```diff
@@ -0,0 +1,219 @@
+
+- Before calling the ledger rule we can query for the needed entries
+- We can present the ledger with the values they asked for, but restricted to
+  the data available in the disk + changelog
+- The ledger will either provide diffs for the given values or will return
+  updated values that we can then diff with the provided ones
+- The resulting differences can be included in the current definition of the
+  `DbChangelog`.
+  
+> **_PLAN:_** Consensus will have a new table on the `LedgerTables` that will
+> represent the unified map. Perhaps even 4 tables or a table of triplets. It
+> will have its own place on the `DbChangelog` too. The flow above describes the
+> general strategy for calling the `BBODY` rule, i.e. `(re)applyBlockOpts` and
+> for calling the `TICK` rule, i.e. `applyTickOpts`.
+
+### Stake distribution
+
+The update to this map is performed by `updateStakeDistribution`. It is known
+before calling the `BBODY` rule which UTxOs are going to be deleted and the rule
+execution logic itself knows the UTxOs that are going to be added. Therefore we
+can follow a logic similar to the above:
+
+- Before calling the ledger rule we can query for the needed entries
+- We can present the ledger with the values they asked for, but restricted to
+  the data available in the disk + changelog
+- The ledger will either provide diffs for the given values or will return
+  updated values that we can then diff with the provided ones
+- The resulting differences can be included in the current definition of the
+  `DbChangelog`.
+  
+> **_PLAN:_** Consensus will have a new table on the `LedgerTables` that will
+> represent the Incremental stake distribution (maybe two tables, one for creds
+> one for ptrs). It will have its own place on the `DbChangelog` too. The flow
+> above describes the general strategy for calling the `BBODY` rule, i.e.
+> `(re)applyBlockOpts`.
+
+### Stake snapshots
+
+The snapshots are rotated by the `snapTransition` rule. This is the most
+complicated of the three because it involves accessing the unified map and the
+`IncrementalStake` in their entirety in order to fold them. We do these two
+steps:
+
+```
+step1 = (dom activeDelegs ◁ credStake) ∪ (dom activeDelegs ◁ ptrStake)
+step2 =  aggregate (dom activeDelegs ◁ rewards) step1
+```
+
+Where 
+- `activeDelegs` comes from the delegations in the unified map,
+- `rewards` come from the unified map,
+- `credStake` and `ptrStake` come from the incremental stake
+
+However, there is an important note here. The `ssStakeMark` is only used to be
+put in `ssStakeSet` on the next snapshot rotation, then `ssStakeSet` is only
+used to be put in `ssStakeGo`, and `ssStakeGo` is only used to prepare the
+`PulsingRewUpdate`. This is done in `PulsingReward.startStep` and it also uses
+the `rewards` map of the unified map. Assuming the reward calculation is done
+outside of the ledger rules, the snapshot is not really needed by the ledger.
```

Comment:
From the preceding discussion, it is unclear how we avoid querying the entire delegations map? Is the fact that `activeDelegs` is a small set the key here?

Reply:
Or is the fact that we're doing the calculation outside the ledger rules the key piece of information here? If so we should probably underline the fact that Ledger HD will require a Consensus-Ledger bridge, and hint at where we think this bridge ought to live (ie Consensus or Ledger).

Reply:
It is the latter, I have clarified it a bit at the bottom.